src/main/java/io/vertx/ext/mongo/impl/GridFSWriteStreamSubscriber.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2018 Red Hat, Inc.
+ *
+ * Red Hat licenses this file to you under the Apache License, version 2.0
+ * (the "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.vertx.ext.mongo.impl;
+
+import io.vertx.core.Promise;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.streams.WriteStream;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.nio.ByteBuffer;
+
+import static io.netty.buffer.Unpooled.copiedBuffer;
+
+
+public class GridFSWriteStreamSubscriber implements Subscriber<ByteBuffer> {
```

Comment:
What was the motivation for this new class? Why not using `PublisherAdapter`, `MappingStream` and `ReadStream#pipeTo`?

Reply:
Example for download to file:

```java
publisher = bucket.downloadToPublisher(fileName);
ReadStream<ByteBuffer> adapter = new PublisherAdapter(context, publisher, 0);
ReadStream<Buffer> rs = new MappingStream( adapter , bb -> Buffer.buffer(copiedBuffer(bb)));
rs.pipeTo(file, promise);
```

Reply:
@tsegismont Indeed it looks cleaner but it breaks compatibility.

GridFSWriteStreamSubscriber can supply the bytes written which is the Future<Long> returned from methods.

I don't know if its useful to anyone but if we go the cleaner way then we have to make the function return Future<Void>

Reply:
GridFS is part of vertx 4 so i assume it should be ok to break the compatibility by changing the methods from Future<Long> to Future<Void> and get rid of GridFSWriteStreamSubscriber

Reply:
Right, then we could create a `CountingStream`  that extends `MappingStream` and accumulates the number of bytes read.
Then when the content is transferred we complete the promise with this value instead of void.

Reply:
@tsegismont it should be ok now